tokens-accounting/src/data.rs

```diff
@@ -0,0 +1,167 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::collections::BTreeMap;
+
+use accounting::{DeltaAmountCollection, DeltaDataCollection, DeltaDataUndoCollection};
+use common::{
+    chain::{
+        tokens::{TokenId, TokenIssuance, TokenTotalSupply},
+        Destination,
+    },
+    primitives::Amount,
+};
+use serialization::{Decode, Encode};
+
+use crate::Error;
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct TokensAccountingData {
+    pub token_data: BTreeMap<TokenId, TokenData>,
+    pub circulating_supply: BTreeMap<TokenId, Amount>,
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct TokensAccountingDeltaData {
+    pub(crate) token_data: DeltaDataCollection<TokenId, TokenData>,
+    pub(crate) circulating_supply: DeltaAmountCollection<TokenId>,
+}
+
+impl TokensAccountingDeltaData {
+    pub fn new() -> Self {
+        Self {
+            token_data: DeltaDataCollection::new(),
+            circulating_supply: DeltaAmountCollection::new(),
+        }
+    }
+
+    pub fn merge_with_delta(
+        &mut self,
+        other: TokensAccountingDeltaData,
+    ) -> Result<TokensAccountingDeltaUndoData, Error> {
+        let token_data_undo = self.token_data.merge_delta_data(other.token_data)?;
+
+        let circulating_supply_undo = other.circulating_supply.clone();
+        self.circulating_supply.merge_delta_amounts(other.circulating_supply)?;
+
+        Ok(TokensAccountingDeltaUndoData {
+            token_data: token_data_undo,
+            circulating_supply: circulating_supply_undo,
+        })
+    }
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct TokensAccountingDeltaUndoData {
+    pub(crate) token_data: DeltaDataUndoCollection<TokenId, TokenData>,
+    pub(crate) circulating_supply: DeltaAmountCollection<TokenId>,
+}
+
+impl TokensAccountingDeltaUndoData {
+    pub fn new() -> Self {
+        Self {
+            token_data: DeltaDataUndoCollection::new(),
+            circulating_supply: DeltaAmountCollection::new(),
+        }
+    }
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub enum TokenData {
+    FungibleToken(FungibleTokenData),
+}
+
+#[derive(Clone, Encode, Decode, Debug, PartialEq, Eq)]
+pub struct FungibleTokenData {
+    token_ticker: Vec<u8>,
+    number_of_decimals: u8,
+    metadata_uri: Vec<u8>,
+    total_supply: TokenTotalSupply,
+    locked: bool,
+    reissuance_controller: Destination,
+}
+
+impl FungibleTokenData {
+    pub fn new(
+        token_ticker: Vec<u8>,
+        number_of_decimals: u8,
+        metadata_uri: Vec<u8>,
+        total_supply: TokenTotalSupply,
+        locked: bool,
+        reissuance_controller: Destination,
+    ) -> Self {
```

Comment:
Would it make sense for the validation of the ticker and uri format and valid values for decimal count to happen here in the constructor? That way we can be sure the fields held by this type are valid.

Alternatively, rename this to `new_unchecked` so it is clear at call sites it is possible to construct illegal token data using this constructor.

Reply:
I'd just raname  it to unchecked because I don't want to complicate it with checks that depend on netupgrade